File: android/app/src/test/kotlin/run/nuri/getagrip/ReminderPlannerTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip

import kotlinx.coroutines.test.runTest
import run.nuri.getagrip.engine.ReminderTime
import run.nuri.getagrip.store.RecordingAlarmScheduler
import run.nuri.getagrip.store.ReminderPlanner
import run.nuri.getagrip.store.ReminderAlarms
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/// The planner is wipe-and-reschedule, so there is no incremental state to corrupt —
/// which only holds if `requests(_)` is a pure function of the routines. These tests
/// exercise that function alone; nothing here touches `AlarmManager`, and the one
/// scheduler that appears is a `RecordingAlarmScheduler` that keeps what it was handed.
class ReminderPlannerTests {

    private val routineA = UUID.fromString("11111111-2222-3333-4444-555555555555")
    private val routineB = UUID.fromString("66666666-7777-8888-9999-000000000000")

    /// `outstanding` defaults high so the plain cases keep asserting the unsuppressed
    /// plan; the suppression cases below set it deliberately.
    private fun input(
        id: UUID,
        name: String,
        times: List<ReminderTime>,
        enabled: Boolean = true,
        outstanding: Int = 99,
    ) = ReminderPlanner.RoutinePlanInput(id, name, times, enabled, outstanding)

    // MARK: - Suppression

    /// The ritual's whole promise is that the app stops asking once you have done it. A
    /// reminder firing after the second session of the day is the app failing to notice
    /// you succeeded — and that is how notifications get switched off for good.
    @Test
    fun slotsAreSuppressedFromTheFrontOfTheDayAsSessionsAreDone() {
        val morning = ReminderTime(hour = 8, minute = 0)
        val evening = ReminderTime(hour = 19, minute = 0)

        val untouched = ReminderPlanner.requests(
            listOf(input(routineA, "Daily", listOf(morning, evening), outstanding = 2))
        )
        assertEquals(2, untouched.size, "nothing done yet, both slots stand")

        // Having trained once, it is the MORNING you have satisfied. Dropping the later
        // slot instead would silence the reminder you still need.
        val halfDone = ReminderPlanner.requests(
            listOf(input(routineA, "Daily", listOf(morning, evening), outstanding = 1))
        )
        assertEquals(2, halfDone.size, "future occurrences must survive today's completion")
        assertEquals(listOf(true, false), halfDone.map { it.suppressToday })

        val finished = ReminderPlanner.requests(
            listOf(input(routineA, "Daily", listOf(morning, evening), outstanding = 0))
        )
        assertEquals(2, finished.size)
        assertTrue(finished.all { it.suppressToday }, "say nothing today, resume tomorrow")
    }

    /// Overshooting the target must not wrap around and re-arm the morning.
    @Test
    fun trainingMoreThanPlannedStillSilencesTheRestOfTheDay() {
        val times = listOf(ReminderTime(hour = 8, minute = 0), ReminderTime(hour = 19, minute = 0))
        val requests = ReminderPlanner.requests(
            listOf(input(routineA, "Daily", times, outstanding = -3))
        )
        assertEquals(2, requests.size)
        assertTrue(requests.all { it.suppressToday })
    }

    @Test
    fun completedDayKeepsTomorrowWithoutReopeningApp() = runTest {
        val zone = ZoneId.of("Europe/Paris")
        val now = LocalDateTime.of(2026, 10, 24, 7, 0)
        val routines = listOf(input(routineA, "Daily", listOf(ReminderTime(hour = 8, minute = 0),
            ReminderTime(hour = 19, minute = 0)), outstanding = 0))
        val scheduler = RecordingAlarmScheduler()
        ReminderPlanner.replan(routines, scheduler)
        assertEquals(2, scheduler.applied.size)
        for (item in scheduler.applied) {
            val first = ReminderAlarms.nextOccurrence(item.hour, item.minute, zone, now, item.suppressToday)
            val expected = now.toLocalDate().plusDays(1).atTime(item.hour, item.minute).atZone(zone)
            assertEquals(expected.toInstant().toEpochMilli(), first)
            // The receiver schedules tomorrow by calendar day, even across DST.
            assertEquals(expected.plusDays(1).toInstant().toEpochMilli(),
                ReminderAlarms.tomorrow(item.hour, item.minute, zone, expected.toLocalDate()))
        }
    }

    /// **A session finished after midnight settles the training day that began at 04:00
    /// YESTERDAY — and must leave tomorrow morning's reminders alone.** The suppression used
    /// to skip a slot's next CALENDAR occurrence, so finishing at 00:30 silenced the 08:00
    /// that belongs to the next training day, and a whole day went unreminded (found on iOS,
    /// and the same arithmetic was here).
    @Test
    fun finishingAfterMidnightDoesNotSilenceTheNextTrainingDay() {
        val zone = ZoneId.of("Europe/Paris")
        val halfPastMidnight = LocalDateTime.of(2026, 9, 23, 0, 30)
        val morning = ReminderAlarms.nextOccurrence(8, 0, zone, halfPastMidnight, suppressToday = true)
        assertEquals(LocalDateTime.of(2026, 9, 23, 8, 0).atZone(zone).toInstant().toEpochMilli(), morning,
            "08:00 today is the next training day's slot, owed in full")

        // A slot still INSIDE the training day being settled is skipped once.
        val lateNight = ReminderAlarms.nextOccurrence(1, 0, zone, halfPastMidnight, suppressToday = true)
        assertEquals(LocalDateTime.of(2026, 9, 24, 1, 0).atZone(zone).toInstant().toEpochMilli(), lateNight)

        // And the ordinary evening case is unchanged: 19:00 today is skipped to tomorrow.
        val sixPm = LocalDateTime.of(2026, 9, 23, 18, 0)
        val evening = ReminderAlarms.nextOccurrence(19, 0, zone, sixPm, suppressToday = true)
        assertEquals(LocalDateTime.of(2026, 9, 24, 19, 0).atZone(zone).toInstant().toEpochMilli(), evening)
        // 04:00 is the NEXT day's first minute, never the settled one's last.
        val beforeRollover = LocalDateTime.of(2026, 9, 23, 3, 0)
        val atRollover = ReminderAlarms.nextOccurrence(4, 0, zone, beforeRollover, suppressToday = true)
        assertEquals(LocalDateTime.of(2026, 9, 23, 4, 0).atZone(zone).toInstant().toEpochMilli(), atRollover)
    }

    /// The front of the TRAINING day: a 01:00 reminder is the evening's last slot, so one
    /// session done suppresses the morning, not the small-hours one.
    @Test
    fun suppressionCountsFromTheFrontOfTheTrainingDay() {
        val late = ReminderTime(hour = 1, minute = 0)
        val morning = ReminderTime(hour = 8, minute = 0)
        val requests = ReminderPlanner.requests(
            listOf(input(routineA, "Daily", listOf(late, morning), outstanding = 1))
        )
        assertEquals(listOf(1, 8), requests.map { it.hour }, "still listed in clock order")
        assertEquals(listOf(false, true), requests.map { it.suppressToday })
    }

    // MARK: - Identity

    /// `ReminderTime` has no UUID: identity IS the time. That makes the notification id
    /// content-keyed, so editing 08:00 → 09:00 replaces the request rather than leaving a
    /// stale 08:00 alarm nobody can see to cancel.
    @Test
    fun identifierIsDeterministicAndContentKeyed() {
        val eight = ReminderTime(hour = 8, minute = 0)
        val nine = ReminderTime(hour = 9, minute = 0)

        assertEquals("r0480", eight.slot)
        assertEquals("r0540", nine.slot)
        assertEquals("r1290", ReminderTime(hour = 21, minute = 30).slot)
        assertEquals("doigt.routine.", ReminderPlanner.identifierPrefix)
        assertEquals(
            "doigt.routine.11111111-2222-3333-4444-555555555555.r0480",
            ReminderPlanner.identifier(routineA, eight),
        )
        // Same inputs, same string, every time — that is what makes a replan idempotent.
        assertEquals(
            ReminderPlanner.identifier(routineA, eight),
            ReminderPlanner.identifier(routineA, eight),
        )

        val before = ReminderPlanner.requests(listOf(input(routineA, "Daily no-hangs", listOf(eight))))
        val after = ReminderPlanner.requests(listOf(input(routineA, "Daily no-hangs", listOf(nine))))
        assertEquals(
            listOf("doigt.routine.11111111-2222-3333-4444-555555555555.r0480"),
            before.map { it.identifier },
        )
        assertEquals(
            listOf("doigt.routine.11111111-2222-3333-4444-555555555555.r0540"),
            after.map { it.identifier },
        )
        assertFalse(
            after.map { it.identifier }.contains(before[0].identifier),
            "the edited slot must not survive alongside its replacement",
        )
    }

    @Test
    fun plannedRequestCarriesTheRoutineNameAndAFiringTime() {
        val planned = ReminderPlanner.requests(
            listOf(input(routineA, "Rest day", listOf(ReminderTime(hour = 19, minute = 15))))
        )
        val request = planned.firstOrNull()
        assertNotNull(request)

        assertEquals("Rest day", request.title)
        assertEquals("Time for a session.", request.body)
        assertEquals(19, request.hour)
        assertEquals(15, request.minute)
    }

    // MARK: - What contributes nothing

    @Test
    fun disabledOrEmptyRoutinesProduceNoRequests() = runTest {
        val times = listOf(ReminderTime(hour = 8, minute = 0), ReminderTime(hour = 19, minute = 0))
        assertEquals(emptyList(), ReminderPlanner.requests(emptyList()))
        assertEquals(
            emptyList(),
            ReminderPlanner.requests(listOf(input(routineA, "Daily no-hangs", times, enabled = false))),
        )
        assertEquals(
            emptyList(),
            ReminderPlanner.requests(listOf(input(routineA, "Daily no-hangs", emptyList()))),
        )

        // A disabled routine must not suppress an enabled one alongside it — and the
        // scheduler is handed exactly what `requests` produced, nothing else.
        val mixed = listOf(
            input(routineA, "Daily no-hangs", times, enabled = false),
            input(routineB, "Rest day", listOf(ReminderTime(hour = 8, minute = 0))),
        )
        assertEquals(listOf("Rest day"), ReminderPlanner.requests(mixed).map { it.title })

        val scheduler = RecordingAlarmScheduler()
        ReminderPlanner.replan(mixed, scheduler)
        assertEquals(listOf("Rest day"), scheduler.applied.map { it.title })
    }

    // MARK: - Sorting and dedupe

    /// Two routines legitimately share a wall-clock time, and the identifier keeps them
    /// apart. What must NOT survive is the same slot twice within one routine — that is
    /// one alarm the user set once, however many times it got into the list.
    @Test
    fun requestsAreSortedAndDeduplicatedAcrossRoutines() {
        val planned = ReminderPlanner.requests(
            listOf(
                input(
                    routineA, "Daily no-hangs",
                    listOf(
                        ReminderTime(hour = 19, minute = 0),
                        ReminderTime(hour = 8, minute = 0),
                        ReminderTime(hour = 8, minute = 0),
                    ),
                ),
                input(routineB, "Rest day", listOf(ReminderTime(hour = 8, minute = 0))),
            )
        )

        assertEquals(3, planned.size)
        assertEquals(3, planned.map { it.identifier }.toSet().size, "identifiers must be unique")
        assertEquals(
            setOf(
                "doigt.routine.11111111-2222-3333-4444-555555555555.r0480",
                "doigt.routine.11111111-2222-3333-4444-555555555555.r1140",
                "doigt.routine.66666666-7777-8888-9999-000000000000.r0480",
            ),
            planned.map { it.identifier }.toSet(),
        )

        // Within a routine the slots come out in clock order regardless of input order.
        assertEquals(
            listOf(8, 19),
            planned.filter { it.title == "Daily no-hangs" }.map { it.hour },
        )
    }

    @Test
    fun reminderTimeClampsToTheDayAndSortsByClock() {
        assertEquals(0, ReminderTime(-30).minutesFromMidnight)
        assertEquals(1439, ReminderTime(5_000).minutesFromMidnight)
        assertEquals(480, ReminderTime(hour = 8, minute = 0).id)
        assertEquals(21, ReminderTime(hour = 21, minute = 30).hour)
        assertEquals(30, ReminderTime(hour = 21, minute = 30).minute)
        assertTrue(ReminderTime(hour = 8, minute = 0) < ReminderTime(hour = 19, minute = 0))

        // The ladder new slots are filled from, in order.
        assertEquals(
            listOf(
                ReminderTime(hour = 8, minute = 0),
                ReminderTime(hour = 19, minute = 0),
                ReminderTime(hour = 12, minute = 30),
                ReminderTime(hour = 21, minute = 30),
            ),
            ReminderTime.defaults,
        )
    }
}

File: Tests/ReminderPlannerTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

/// The planner is wipe-and-reschedule, so there is no incremental state to corrupt —
/// which only holds if `requests(for:)` is a pure function of the routines. These tests
/// exercise that function alone; nothing here touches UNUserNotificationCenter.
@MainActor
final class ReminderPlannerTests: XCTestCase {

    private let routineA = UUID(uuidString: "11111111-2222-3333-4444-555555555555")!
    private let routineB = UUID(uuidString: "66666666-7777-8888-9999-000000000000")!

    private func input(_ id: UUID, _ name: String, _ times: [ReminderTime],
                       enabled: Bool = true,
                       outstanding: Int = 99) -> ReminderPlanner.RoutinePlanInput {
        // `outstanding` defaults high so the pre-existing tests keep asserting the
        // unsuppressed plan; the suppression cases below set it deliberately.
        ReminderPlanner.RoutinePlanInput(id: id, name: name, reminders: times,
                                         enabled: enabled, outstandingToday: outstanding)
    }

    // MARK: - Suppression

    /// The ritual's whole promise is that the app stops asking once you have done it.
    /// A reminder firing after the second session of the day is the app failing to
    /// notice you succeeded — and that is how notifications get switched off for good.
    func testSlotsAreSuppressedFromTheFrontOfTheDayAsSessionsAreDone() {
        let morning = ReminderTime(hour: 8, minute: 0)
        let evening = ReminderTime(hour: 19, minute: 0)

        let untouched = ReminderPlanner.requests(
            for: [input(routineA, "Daily", [morning, evening], outstanding: 2)])
        XCTAssertEqual(untouched.count, 2, "nothing done yet, both slots stand")

        // Having trained once, it is the MORNING you have satisfied. Dropping the later
        // slot instead would silence the reminder you still need.
        let halfDone = ReminderPlanner.requests(
            for: [input(routineA, "Daily", [morning, evening], outstanding: 1)])
        XCTAssertEqual(halfDone.count, 2, "future occurrences must survive today's completion")
        XCTAssertEqual(halfDone.map(\.suppressToday), [true, false])

        let finished = ReminderPlanner.requests(
            for: [input(routineA, "Daily", [morning, evening], outstanding: 0)])
        XCTAssertEqual(finished.count, 2)
        XCTAssertTrue(finished.allSatisfy(\.suppressToday), "say nothing today, resume tomorrow")
    }

    /// Overshooting the target must not wrap around and re-arm the morning.
    func testTrainingMoreThanPlannedStillSilencesTheRestOfTheDay() {
        let times = [ReminderTime(hour: 8, minute: 0), ReminderTime(hour: 19, minute: 0)]
        let requests = ReminderPlanner.requests(
            for: [input(routineA, "Daily", times, outstanding: -3)])
        XCTAssertEqual(requests.count, 2)
        XCTAssertTrue(requests.allSatisfy(\.suppressToday))
    }

    func testCompletedDayKeepsTomorrowAndFillsOnlyAvailableBudget() throws {
        let calendar = DayStamp.utcCalendar
        let now = try XCTUnwrap(calendar.date(from: DateComponents(year: 2026, month: 9, day: 6, hour: 7)))
        let routines = [input(routineA, "Daily", [ReminderTime(hour: 8, minute: 0),
                                                 ReminderTime(hour: 19, minute: 0)], outstanding: 0)]
        let planned = ReminderPlanner.scheduledRequests(for: routines, now: now, calendar: calendar, limit: 3)
        let dates = try planned.map { try XCTUnwrap(calendar.date(from: $0.components)) }
        XCTAssertEqual(dates.map { DayStamp(date: $0, calendar: calendar) }, [
            DayStamp(year: 2026, month: 9, day: 7), DayStamp(year: 2026, month: 9, day: 7),
            DayStamp(year: 2026, month: 9, day: 8)])
        XCTAssertEqual(planned.map { $0.components.hour }, [8, 19, 8])
        XCTAssertEqual(Set(planned.map(\.identifier)).count, 3)
        XCTAssertTrue(ReminderPlanner.scheduledRequests(for: routines, now: now,
                                                       calendar: calendar, limit: 0).isEmpty)
    }

    func testDatedHorizonIsChronologicalAcrossRoutinesAndSkipsPastSlots() throws {
        let calendar = DayStamp.utcCalendar
        let now = try XCTUnwrap(calendar.date(from: DateComponents(year: 2026, month: 9, day: 6, hour: 10)))
        let routines = [input(routineA, "Evening", [ReminderTime(hour: 19, minute: 0)]),
                        input(routineB, "Morning", [ReminderTime(hour: 8, minute: 0)])]
        let planned = ReminderPlanner.scheduledRequests(for: routines, now: now, calendar: calendar, limit: 4)
        XCTAssertEqual(planned.map(\.title), ["Evening", "Morning", "Evening", "Morning"])
        let replanned = ReminderPlanner.scheduledRequests(for: routines, now: now,
                                                         calendar: calendar, limit: 4)
        XCTAssertEqual(planned, replanned, "replanning must replace the same dated IDs")
    }

    // MARK: - The training day, not the calendar day

    /// The dates the plan fires at, in `calendar`, as (day of month, hour, minute).
    private func firings(_ planned: [ReminderPlanner.PlannedReminder],
                         _ calendar: Calendar) throws -> [[Int]] {
        try planned.map {
            let date = try XCTUnwrap(calendar.date(from: $0.components))
            let c = calendar.dateComponents([.day, .hour, .minute], from: date)
            return [c.day!, c.hour!, c.minute!]
        }
    }

    /// Finishing the day's last session at 00:15 satisfied the EVENING BEFORE — that is
    /// the training day `outstandingToday` is counted for. Suppressing by calendar day
    /// silenced the new morning's 08:00 and 19:00 instead, and the replan then retired
    /// them from the notification center.
    func testASessionAfterMidnightDoesNotSilenceTheComingDay() throws {
        let calendar = DayStamp.utcCalendar
        let now = try XCTUnwrap(calendar.date(from: DateComponents(year: 2026, month: 9, day: 21,
                                                                   hour: 0, minute: 30)))
        let routines = [input(routineA, "Daily", [ReminderTime(hour: 8, minute: 0),
                                                 ReminderTime(hour: 19, minute: 0)], outstanding: 0)]
        let planned = ReminderPlanner.scheduledRequests(for: routines, now: now,
                                                        calendar: calendar, limit: 3)
        XCTAssertEqual(try firings(planned, calendar), [[21, 8, 0], [21, 19, 0], [22, 8, 0]])
    }

    /// At 23:30 the training day still has until 04:00 to run, so a finished day
    /// silences a 01:00 slot on the next calendar date too — and nothing past 04:00.
    func testAFinishedEveningAlsoSilencesItsSmallHoursSlot() throws {
        let calendar = DayStamp.utcCalendar
        let now = try XCTUnwrap(calendar.date(from: DateComponents(year: 2026, month: 9, day: 20,
                                                                   hour: 23, minute: 30)))
        let routines = [input(routineA, "Late", [ReminderTime(hour: 23, minute: 45),
                                                ReminderTime(hour: 1, minute: 0),
                                                ReminderTime(hour: 8, minute: 0)], outstanding: 0)]
        let planned = ReminderPlanner.scheduledRequests(for: routines, now: now,
                                                        calendar: calendar, limit: 4)
        XCTAssertEqual(try firings(planned, calendar),
                       [[21, 8, 0], [21, 23, 45], [22, 1, 0], [22, 8, 0]])

        // Nothing done yet: the same evening keeps every slot it still has.
        let owed = [input(routineA, "Late", [ReminderTime(hour: 23, minute: 45),
                                            ReminderTime(hour: 1, minute: 0)], outstanding: 2)]
        XCTAssertEqual(try firings(ReminderPlanner.scheduledRequests(for: owed, now: now,
                                                                     calendar: calendar, limit: 2),
                                   calendar),
                       [[20, 23, 45], [21, 1, 0]])
    }

    /// Past the rollover it is a new training day: one session of two done at 04:15
    /// silences that morning and leaves the evening owed.
    func testAfterTheRolloverTheMorningIsTheOneSatisfied() throws {
        let calendar = DayStamp.utcCalendar
        let now = try XCTUnwrap(calendar.date(from: DateComponents(year: 2026, month: 9, day: 21,
                                                                   hour: 4, minute: 30)))
        let routines = [input(routineA, "Daily", [ReminderTime(hour: 8, minute: 0),
                                                 ReminderTime(hour: 19, minute: 0)], outstanding: 1)]
        let planned = ReminderPlanner.scheduledRequests(for: routines, now: now,
                                                        calendar: calendar, limit: 2)
        XCTAssertEqual(try firings(planned, calendar), [[21, 19, 0], [22, 8, 0]])
    }

    /// "The front of the day" is the front of the TRAINING day: with one of two done,
    /// the 08:00 is the satisfied one and the 01:00 — the end of that evening — is owed.
    func testSuppressionCountsFromTheRolloverNotFromMidnight() {
        let requests = ReminderPlanner.requests(
            for: [input(routineA, "Owl", [ReminderTime(hour: 1, minute: 0),
                                          ReminderTime(hour: 8, minute: 0)], outstanding: 1)])
        XCTAssertEqual(requests.map { $0.components.hour }, [8, 1])
        XCTAssertEqual(requests.map(\.suppressToday), [true, false])
    }

    // MARK: - Identity

    /// `ReminderTime` has no UUID: identity IS the time. That makes the notification id
    /// content-keyed, so editing 08:00 → 09:00 replaces the request rather than leaving
    /// a stale 08:00 alarm nobody can see to cancel.
    func testIdentifierIsDeterministicAndContentKeyed() {
        let eight = ReminderTime(hour: 8, minute: 0)
        let nine = ReminderTime(hour: 9, minute: 0)

        XCTAssertEqual(eight.slot, "r0480")
        XCTAssertEqual(nine.slot, "r0540")
        XCTAssertEqual(ReminderTime(hour: 21, minute: 30).slot, "r1290")
        XCTAssertEqual(ReminderPlanner.identifierPrefix, "doigt.routine.")
        XCTAssertEqual(ReminderPlanner.identifier(routine: routineA, slot: eight),
                       "doigt.routine.11111111-2222-3333-4444-555555555555.r0480")
        // Same inputs, same string, every time — that is what makes a replan idempotent.
        XCTAssertEqual(ReminderPlanner.identifier(routine: routineA, slot: eight),
                       ReminderPlanner.identifier(routine: routineA, slot: eight))

        let before = ReminderPlanner.requests(for: [input(routineA, "Daily no-hangs", [eight])])
        let after = ReminderPlanner.requests(for: [input(routineA, "Daily no-hangs", [nine])])
        XCTAssertEqual(before.map(\.identifier), ["doigt.routine.11111111-2222-3333-4444-555555555555.r0480"])
        XCTAssertEqual(after.map(\.identifier), ["doigt.routine.11111111-2222-3333-4444-555555555555.r0540"])
        XCTAssertFalse(after.map(\.identifier).contains(before[0].identifier),
                       "the edited slot must not survive alongside its replacement")
    }

    func testPlannedRequestCarriesTheRoutineNameAndAFiringTime() throws {
        let planned = ReminderPlanner.requests(
            for: [input(routineA, "Rest day", [ReminderTime(hour: 19, minute: 15)])])
        let request = try XCTUnwrap(planned.first)

        XCTAssertEqual(request.title, "Rest day")
        XCTAssertEqual(request.body, "Time for a session.")
        XCTAssertEqual(request.components.hour, 19)
        XCTAssertEqual(request.components.minute, 15)
    }

    // MARK: - What contributes nothing

    func testDisabledOrEmptyRoutinesProduceNoRequests() {
        let times = [ReminderTime(hour: 8, minute: 0), ReminderTime(hour: 19, minute: 0)]
        XCTAssertEqual(ReminderPlanner.requests(for: []), [])
        XCTAssertEqual(ReminderPlanner.requests(for: [input(routineA, "Daily no-hangs", times, enabled: false)]), [])
        XCTAssertEqual(ReminderPlanner.requests(for: [input(routineA, "Daily no-hangs", [])]), [])

        // A disabled routine must not suppress an enabled one alongside it.
        let mixed = ReminderPlanner.requests(for: [
            input(routineA, "Daily no-hangs", times, enabled: false),
            input(routineB, "Rest day", [ReminderTime(hour: 8, minute: 0)]),
        ])
        XCTAssertEqual(mixed.map(\.title), ["Rest day"])
    }

    // MARK: - Sorting and dedupe

    /// Two routines legitimately share a wall-clock time, and the identifier keeps them
    /// apart. What must NOT survive is the same slot twice within one routine — that is
    /// one alarm the user set once, however many times it got into the array.
    func testRequestsAreSortedAndDeduplicatedAcrossRoutines() {
        let planned = ReminderPlanner.requests(for: [
            input(routineA, "Daily no-hangs", [ReminderTime(hour: 19, minute: 0),
                                               ReminderTime(hour: 8, minute: 0),
                                               ReminderTime(hour: 8, minute: 0)]),
            input(routineB, "Rest day", [ReminderTime(hour: 8, minute: 0)]),
        ])

        XCTAssertEqual(planned.count, 3)
        XCTAssertEqual(Set(planned.map(\.identifier)).count, 3, "identifiers must be unique")
        XCTAssertEqual(Set(planned.map(\.identifier)), [
            "doigt.routine.11111111-2222-3333-4444-555555555555.r0480",
            "doigt.routine.11111111-2222-3333-4444-555555555555.r1140",
            "doigt.routine.66666666-7777-8888-9999-000000000000.r0480",
        ])

        // Within a routine the slots come out in clock order regardless of input order.
        let aSlots = planned.filter { $0.title == "Daily no-hangs" }.map { $0.components.hour }
        XCTAssertEqual(aSlots, [8, 19])
    }

    func testReminderTimeClampsToTheDayAndSortsByClock() {
        XCTAssertEqual(ReminderTime(minutesFromMidnight: -30).minutesFromMidnight, 0)
        XCTAssertEqual(ReminderTime(minutesFromMidnight: 5_000).minutesFromMidnight, 1439)
        XCTAssertEqual(ReminderTime(hour: 8, minute: 0).id, 480)
        XCTAssertEqual(ReminderTime(hour: 21, minute: 30).hour, 21)
        XCTAssertEqual(ReminderTime(hour: 21, minute: 30).minute, 30)
        XCTAssertLessThan(ReminderTime(hour: 8, minute: 0), ReminderTime(hour: 19, minute: 0))

        // The ladder new slots are filled from, in order.
        XCTAssertEqual(ReminderTime.defaults, [ReminderTime(hour: 8, minute: 0),
                                               ReminderTime(hour: 19, minute: 0),
                                               ReminderTime(hour: 12, minute: 30),
                                               ReminderTime(hour: 21, minute: 30)])
    }
}
